core/container.go

```diff
@@ -289,30 +289,59 @@ func (container *Container) From(ctx context.Context, addr string) (*Container,
 		return nil, fmt.Errorf("failed to get buildkit client: %w", err)
 	}
 
-	container = container.Clone()
-
 	platform := container.Platform
 
 	refName, err := reference.ParseNormalizedNamed(addr)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to parse image address %s: %w", addr, err)
 	}
+	// add a default :latest if no tag or digest, otherwise this is a no-op
+	refName = reference.TagNameOnly(refName)
 
-	ref := reference.TagNameOnly(refName).String()
+	if refName, isCanonical := refName.(reference.Canonical); isCanonical {
+		return container.FromCanonical(ctx, refName)
+	}
 
-	_, digest, cfgBytes, err := bk.ResolveImageConfig(ctx, ref, sourceresolver.Opt{
+	_, digest, _, err := bk.ResolveImageConfig(ctx, refName.String(), sourceresolver.Opt{
 		Platform: ptr(platform.Spec()),
 		ImageOpt: &sourceresolver.ResolveImageOpt{
 			ResolveMode: llb.ResolveModeDefault.String(),
 		},
 	})
 	if err != nil {
-		return nil, fmt.Errorf("failed to resolve image %s: %w", ref, err)
+		return nil, fmt.Errorf("failed to resolve image %s: %w", refName.String(), err)
+	}
+	canonRefName, err := reference.WithDigest(refName, digest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set digest on image %s: %w", refName.String(), err)
 	}
 
-	digested, err := reference.WithDigest(refName, digest)
+	return container.FromCanonical(ctx, canonRefName)
+}
+
+func (container *Container) FromCanonical(ctx context.Context, refName reference.Canonical) (*Container, error) {
```

Comment:
nit: should `container.From` take a `reference.Reference` to match this?

Reply:
That just puts the onus on the caller to do more stuff itself and handle more errors, etc. so don't think it's worth it. I renamed the methods to be `FromRefString` and `FromCanonicalRef` to clear it up more though.